### PR TITLE
Tests for blurredLocation.getPlacenameFromCoordinates()

### LINKED ISSUE #24: Change getPlacenameFromCoordinates() to be asynchronous and use a callback handler
`getPlacenameFromCoordinates()` is not working because it makes an AJAX call and that runs asynchronously:

```js
  function getPlacenameFromCoordinates(lat, lng) {
    var loc = "Location not found";

    $.getJSON("https://maps.googleapis.com/maps/api/geocode/json?latlng="+lat+","+lng, function(data) {
      if (data.results[0]) {
        loc = data.results[0].formatted_address;
      }
    });
    return loc;
  }
```

https://github.com/publiclab/leaflet-blurred-location/blob/master/src/core/object.js#L78-L87

At the time it wants to return `loc` the response hasn't come back yet. So we need to pass it a function to be run on completion, and you can tell it what to do with the location string there:

`getPlacenameFromCoordinates(lat, lon, callback)`

the callback can be anything, even `alert()` -- so:

`getPlacenameFromCoordinates(10, 40, alert)` would pop up an alert with the result.

### PR DESCRIPTION
**Pull request by mridulnagpal**
closes #24

### CODE REVIEW

File: Gruntfile.js [hunk 1]
```diff
@@ -22,7 +22,7 @@ module.exports = function(grunt) {
 
         browserify: {
             dist: {
-                src: ['src/core/object.js'],
+                src: ['node_modules/jquery/dist/jquery.min.js', 'node_modules/leaflet/dist/leaflet.js', 'src/blurredLocation.js'],
```
Comment (jywarren) [inline Gruntfile.js]: Oops be sure to work from the latest master!

Comment (mridulnagpal) [inline Gruntfile.js]: I took it from my restruct PR so that there won't be any conflicts again

**Comment by mridulnagpal** [comment]
Is it ready to merge?

**Comment by mridulnagpal** [comment]
@jywarren Ready to merge?

**Comment by jywarren** [comment]
Not quite -- the async version isn't too hard -- take a look:

```js
function getPlacenameFromCoordinates(lat, lng, onResponse) {
  $.ajax({
    url:"https://maps.googleapis.com/maps/api/geocode/json?latlng=" + lat + "," + lng,
    async: false,
    success: function(result) {
      onResponse(result.results[0].formatted_address);
    }
  });
  // here, don't need to return anything because it won't be ready yet!
}
```

Then, to use it, you have to pass it a function to run when it's done:

```js
getPlacenameFromCoordinates(51, 62, function onResponse(result) {
  console.log(result);
})
```

There's this really great book on JavaScript here: https://ponyfoo.com/books/javascript-application-design -- i don't know if you can get a copy, but there are also lots of good blog posts on it too -- take a look http://recurial.com/programming/understanding-callback-functions-in-javascript/

**Comment by mridulnagpal** [comment]
@jywarren Sounds cool will look into it

**Comment by mridulnagpal** [comment]
@jywarren I made the changes but what if I want the function to return the value of result, for example something like `var temp = blurredLocation.getPlacenameFromCoordinates(51, 62)` and temp has the result now. Which function do I pass as onResponse if I want this to happen

**Comment by jywarren** [comment]
It depends on what you want to do next. Basically you can put code in the
function that'll be run after the response is available. So if you want to
display it in html, you can put the code to display it in the callback
function.

**Comment by mridulnagpal** [comment]
Got it, but how do we write a test for it?

**Comment by jywarren** [comment]
Great question. There is actually a jasmine test type for asynchronous
functions. It also uses a callback.

This is the documentation for v1.3 but it's better than the 2.x docs --

https://jasmine.github.io/1.3/introduction#section-Asynchronous_Support

**Comment by mridulnagpal** [comment]
Looking into it, until then please have a look at the new PR added for setPrecision()

**Comment by mridulnagpal** [comment]
Added test spec please have a look

**Comment by mridulnagpal** [comment]
@jywarren Do you think this is ready to merge?

**Comment by jywarren** [comment]
OK - i left a suggestion in a sub-PR against this branch: https://github.com/mridulnagpal/leaflet-blurred-location/pull/2/files

If it looks good, merge it into here, and then this one will update.

**Comment by mridulnagpal** [comment]
The test works indeed, thanks :smile:

**Comment by mridulnagpal** [comment]
@jywarren This worked in my remote repo

**Comment by jywarren** [comment]
Is it because the Travis system can't make remote calls? Maybe we have to mock up the response. That could be a little complex:

https://github.com/jasmine/jasmine-ajax

But you can see how it works here: https://github.com/publiclab/spectral-workbench.js/blob/master/spec/javascripts/api_spec.js

Basically you make a fake response that the test will get back, and it'll think the Google API has responded. That way it's deterministic, too.

**Comment by jywarren** [comment]
See how it doesn't quite make sense to be running an automated test on something that depends on a remote server response. But this is a way around that.

**Comment by jywarren** [comment]
We could refactor this to just include the `+    blurredLocation.getPlacenameFromCoordinates()` test, and mock an ajax response, but it's not high priority as we're essentially wrapping an external library's function pretty simply without providing new functionality.

**Comment by mridulnagpal** [comment]
Cool, I'll look into it.

**Comment by mridulnagpal** [comment]
Made the function return a fake response for the test. Please have a look.

**Comment by jywarren** [comment]
This is a good use of a spy, but you're not actually testing the method, because you're replacing it with the spy method. I think you have to look at this example for how to fake a complete Ajax response from a remote server: https://github.com/publiclab/PublicLab.Editor/blob/master/spec/javascripts/editor_spec.js#L45-L76

**Comment by jywarren** [comment]
But I would say that this is lower priority than the Graticule work in #36 -- so you don't have to worry to much about this now! It's a pretty simple method we don't absolutely need a test for.

**Comment by mridulnagpal** [comment]
Let's close this for now, will rebase and continue later

**Pull request closed without merging**